docs/developer-docs/security/rust-canister-development-security-best-practices.mdx

```diff
@@ -831,6 +831,116 @@ Publicly revealing the canister's cycles balance allows an attacker to measure t
 
 Your canisters should not publicly expose their cycles balance (available through the system API), i.e., they should only expose their cycles balance to their controllers or other trusted principals.
 
+### Using secure randomness in canisters
+
+Canister developers often require access to secure randomness in their canisters to perform certain operations. The requirements for a secure randomness source include 
+* **Unbiased:** The value shouldn't be influenced by anyone.
+* **Unpredictable:** The value is unknown to anyone before it is generated.
+
+The Internet Computer exposes the system API [`raw_rand`](https://internetcomputer.org/docs/current/references/ic-interface-spec#ic-raw_rand) for this exact purpose, which accepts no input and returns 32 bytes of cryptographically secure randomness. The secure randomness guarantees of `raw_rand` are achieved by the cryptographic properties of the [RandomTape](https://internetcomputer.org/how-it-works/execution-layer/#random-number-generation). It is always recommended to use `raw_rand` as a source of randomness in canisters and **avoid** using other sources with low entropy such as current time.
+
+To illustrate the usage of `raw_rand`, we discuss two examples in Motoko and Rust below, while weighing the benefits and caveats around using them.
+
+#### 1. Direct usage of `raw_rand` as the random number generator
+
+In this Motoko example, the canister provides the requested size of random bytes by calling `raw_rand`. However, it can only generate 32 bytes of secure randomness in a single message and thus subsequent calls to the system API is required to fill the requested size.
+
+```motoko
+import Random "mo:base/Random";
+import Array "mo:base/Array";
+
+actor Randomness {
+  public func random_bytes(n : Nat) : async [Nat8] {
+    let byteArray : [var Nat8] = Array.init<Nat8>(n, 0);
+    let entropy = await Random.blob();
+    var f = Random.Finite(entropy);
+    var i = 0;
+    loop {
+      if (i == n) {
+        return Array.freeze<Nat8>(byteArray);
+      } else {
+        switch (f.byte()) {
+          case (?byte) {
+            byteArray[i] := byte;
+            i := i + 1;
+          };
+          case null {
+            let entropy = await Random.blob();
+            f := Random.Finite(entropy);
+          };
+        };
+      };
+    };
+  };
+};
+```
+
+##### Benefits:
+- The random bytes is guaranteed to be secure.
+- The method can be used in a query as it doesn't store any state.
```

Comment:
Yeah, this can't be used in a query. I was under the assumption we used the random tape of previous round for RNG and thus no state change is needed. 

I removed the line.